text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+(The number of prerequisites of a Work Package is limited to at most one. However, we cannot trivially control the number of dependents in the same way, nor would we necessarily wish to since it would open up a griefing vector for misbehaving Work Package Builders who interrupt a sequence by introducing their own Work Packages with a prerequisite which is within another's sequence.)
+
+Work Items are a pair of class and payload, where the `class` identifies the Class of Work to be done in this item (*Work Class*).
+
+Though this process happens entirely in consensus, there are two main consensus environments at play, _in-core_ and _on-chain_. We therefore partition the progress into two pairs of stages: Collect & Refine and Join & Accumulate.
+
+### Processing stages of a Work Package
+
+A Work Package has several stages of consensus computation associated with its processing, which happen as the system becomes more certain that it represents a correct and useful transition of its Work Class.
+
+While a Work Package is being built, the *Builder* must have a synchronized Relay-chain node in order to supply a specific *Context*. The Context dictates a certain *Scope* for the Work Package which is used by the Basic Validation to limit which Relay-chain blocks it may be processed on to a small sequence of a specific fork (which is yet to be built, presumably). We define the Relay-chain height at this point to be `T`.
+
+The first consensus computation to be done is the Work Package having its Authorization checked in-core, hosted by the Relay-chain Validator Group. If it is determined to be authorized, then the same environment hosts the Refinement of the Work Package into a series of Work Results. This concludes the bulk of the computation that the Work Package represents. We would assume that the Relay-chain's height at this point is shortly after the authoring time, `T+r` where `r` could be as low as zero.
+
+The second consensus computation happens on-chain at the behest of the Relay-chain Block Author of the time `T+r+i`, where `i` is generally zero or one, the time taken for the Work Results to be transported from within the Core to get to the gateway of being on-chain. The computation done essentially just ensures that the Work Package is still in scope and that the prerequisite it relies upon (if any) has been submitted ahead of it. This is called the on-chain *Reporting* (in the fixed-function parachains model, this is known as "attestation") and initiates the *Availability Protocol* for this Work Package once Relay-chain Validators synchronize to the block. This protocol guarantees that the Work Package will be made available for as long as we allow disputes over its validity to be made.
```

Comment:
I thought disputes was the reason we have this ultra-secure availability system, no?

Reply:
We can not revert past the last finalized block and under normal operations we have finality in 2 - 3 blocks. Disputes can only happen for non finalized blocks. However, we store all PoVs on the canonical chain for 24 hours.

Reply:
Sure, but isn't that over-conservative for the purpose they are serving (i.e. ensuring availability in case of dispute)?

Reply:
Finality can stall up to dispute_period, which is currently configured to 6 sessions on Polkadot, so yes 24 hours. Although availability period is 25 hours. 
However, availability is also useful for parachain liveness. In case, there's a malicious collator that produced a block but didn't reveal it to other collators, so that can't build blocks on top. In that case, they can recover the state changes from the relay chain. Here a longer period might be desired.

Reply:
> However, availability is also useful for parachain liveness.

But is that really what it's designed for? Can all parachains (if we scale to 10s of 1000s via ODP) realistically request 24 hours after finality? I was under the impression this was only an afterthought and not actually sound.

Reply:
Preserving liveness by protecting against the class of Block Withholding attacks is a key goal of the data availability system. The system is designed for this, although it can become a bottleneck when it is being used as a first resort rather than a last resort in the case that the data is not generally available over less concentrated p2p networks.

> this doesn't seem so feasible since new validators would need to sync an entire month's worth of erasure-coded fragments when they come online.

It's rather that validators are expected to stay online for the availability period - there are no handovers of availability erasure-coded chunks in the current protocol.

Reply:
Yes, we gain liveness from availability.  We gain some liveness even if we lack code to handle this.

Also, it helps governance resolve post finality disputes like what kusama experienced too.  We should never have those of course, so it might also make sense to have parachains whose erasure coded chunks were pruned more eagerly, but this looks unclear right now.

Aside from full PoVs, availability looks useful in messages too.  We'd never want to rerun whole PoVs to obtain messages, so the messages could/should always be erasure coded separately:  Output style outgoing messages must be erasure coded separately from the PoV.  Input style outgoing messages could be erasure coded seperately from the PoV.   In both cases, we'd love to pass them pre-encoded as the same erasure code to the receiving chain's parablock.  Also multi-block polkavm sessions become weird here.  In other words, we've a variety of erasure code patterns which shall arise in the future.

Reply:
> Preserving liveness by protecting against the class of Block Withholding attacks is a key goal of the data availability system. The system is designed for this, although it can become a bottleneck when it is being used as a first resort rather than a last resort in the case that the data is not generally available over less concentrated p2p networks.
> 
> > this doesn't seem so feasible since new validators would need to sync an entire month's worth of erasure-coded fragments when they come online.
> 
> It's rather that validators are expected to stay online for the availability period - there are no handovers of availability erasure-coded chunks in the current protocol.

Without handovers, it seems that the minimum period a validator may be elected without being slashed must be pretty much equivalent to the maximum period by which data is available.

Reply:
Has any thought been given to a secondary available subsystem specifically to enable longer periods of storage?

Reply:
Validator bonds are held for 28 days and the validator can be slashed at any point, but availability withholding is not a slashable offence anyway.

Reply:
I meant "the minimum period a validator can be elected for, disregarding the possibility of them being kicked out due to misbehaviour".

Reply:
Yes.  We've my old XCM transport documents, and I'll do an "off-chain transport" pre-RFC eventually, but in brief..

We erasure code for 2/3rds of a fixed number of validators.  We do not have a fixed number of validators over a longer time scale, but validators do know who held what pieces earlier, including who acted as approval checkers.  We'd export this information via an interface morally similar to bittorrent trackers, except each validator has differing possibly malicious views, so clients should randomly ask for second opinions.

Ideally, you fetch systemic chunks from approval checkers because this (a) avoids reconstruction and (b) approval checkers are numerous enough to obstruct censorship.  A tracker-like interface permits doing this, without exposing the relay chain's complexity to parachains.